sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -10,109 +10,75 @@
         {
             if (message.Request.Uri.Scheme != Uri.UriSchemeHttps)
             {
                 throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS protected (https) endpoints.");
             }
 
-            RequestContent originalContent = message.Request.Content;
-
-            // if this policy doesn't have _challenge cached try to get it from the static challenge cache
+            // if this policy doesn't have _challenge cached try to get it from the static challenge cache.
             AuthenticationChallenge challenge = _challenge ?? AuthenticationChallenge.GetChallenge(message);
 
-            // if we still don't have the challenge for the endpoint
-            // remove the content from the request and send without authentication to get the challenge
+            // if we don't have the challenge for the endpoint, remove the content from the request and send without authentication to get the challenge
             if (challenge == null)
             {
+                RequestContent originalContent = message.Request.Content;
                 message.Request.Content = null;
-            }
-            // otherwise if we already know the challenge authenticate the request
-            else
-            {
-                await AuthenticateRequestAsync(message, async, challenge).ConfigureAwait(false);
-            }
-
-            if (async)
-            {
-                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
-            }
-            else
-            {
-                ProcessNext(message, pipeline);
-            }
 
-            // if we get a 401
-            if (message.Response.Status == 401)
-            {
-                // set the content to the original content in case it was cleared
-                message.Request.Content = originalContent;
-
-                // update the cached challenge
-                challenge = AuthenticationChallenge.GetChallenge(message);
-
-                if (challenge != null)
+                if (async)
                 {
-                    // update the cached challenge if not yet set or different from the current challenge (e.g. moved tenants)
-                    if (_challenge == null || !challenge.Equals(_challenge))
-                    {
-                        _challenge = challenge;
-                    }
-
-                    // authenticate the request and resend
-                    await AuthenticateRequestAsync(message, async, challenge).ConfigureAwait(false);
-
-                    if (async)
-                    {
-                        await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        ProcessNext(message, pipeline);
-                    }
+                    await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
```

Comment:
If this is only `PreProcessAsync` and `Process(Async)` in the base does this, why do it here? In fact, why is this pre- and post-processing necessary? It seems the only difference is that CAE uses `claims` and different headers, but flow is effectively the same.

Reply:
Only Key Vault sends an initial empty request to receive the first challenge. I assumed this specific behavior was required. If not, the pre-processing is not necessary. 

There is technically not any post processing, just a resumption of normal request flow. For example, once we've cached the initial scope, the pre-processing is skipped on subsequent requests.

Reply:
Not required, but Identity doesn't have a way to pass the tenant ID along just yet. /cc @schaabs who could explain more, if necessary.

Reply:
I'll change my vote since this at least answers my question, but still feels a bit odd that the class is inherited by doesn't seem to actually inherit that much functionality. While I'm 100% in favor of sharing common features like this, it just doesn't seem like much is actually shared to the point of: maybe it's not worth it right now?

Reply:
OK, you motivated me to try refactoring more into the base 😄 . The latest commit eliminates the logic of GetChallenge from  `ChallengeBasedAuthenticationPolicy`. It also gets faster with less allocations.

```
|      Method |     Mean |   Error |  StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------:|--------:|--------:|------:|--------:|-------:|------:|------:|----------:|
|    KeyVault | 599.1 ns | 3.28 ns | 2.74 ns |  1.59 |    0.02 | 0.1631 |     - |     - |    1368 B |
| KeyVaultNew | 377.7 ns | 4.78 ns | 3.99 ns |  1.00 |    0.00 | 0.0172 |     - |     - |     144 B |
```

Reply:
> Only Key Vault sends an initial empty request to receive the first challenge

The body is removed from the request because Key Vault supports other authentication schemes which also protect the body of the request, so before we know the auth scheme we need to avoid sending an unprotected body to key vault.  We don't currently support this enhanced auth scheme in the SDK but we still don't want to send any unprotected data to vaults which require it.